### ISSUE TITLE
express.d.ts - overload is not recognized

### ISSUE THREAD

**Issue by DmitryEfimenko** [borisyankov/DefinitelyTyped#4212]
The following code reports an issue:
```javascript
import express = require('express');
var app = express();

if (app.get('env') === 'development') {
    app.use(function (err, req, res, next) {
        res.status(err.status || 500);
        res.render('error', {
            message: err.message,
            error: err
        });
    });
}
```
It's supposed to be using the following overload
```javascript
use(path: string, handler: ErrorRequestHandler): T;
```

**Comment by vvakame** [comment]
@DmitryEfimenko please send a PR. I'll review it.

**Comment by DmitryEfimenko** [comment]
well, I don't know what the issue is with the `express.d.ts`. Everything seems to be in order and yet:
![_ts_error](https://cloud.githubusercontent.com/assets/2098175/7385458/978eaca8-edfb-11e4-9bfa-1c47ca66be7d.png)

FYI: it happens in Visual Studio 2013.

**Comment by DmitryEfimenko** [comment]
the way I worked around it (without fixing the definitions):
```javascript
var errorHandlerDev: express.ErrorRequestHandler = function (err, req, res, next) {
    res.status(err.status || 500);
    res.render('error', {
        message: err.message,
        error: err
    });
};

if (app.get('env') === 'development') {
    app.use(errorHandlerDev);
}
```

**Comment by ScottReed** [comment]
I'm getting the same issue is Visual Studio Code, thanks for the workaround @DmitryEfimenko

**Comment by cwmacdon** [comment]
```
import express = require('express');
var app = express();

if (app.get('env') === 'development') {
    app.use(<express.ErrorRequestHandler>function (err, req, res, next) {
        res.status(err.status || 500);
        res.render('error', {
            message: err.message,
            error: err
        });
    });
}
```

**Comment by kunl** [comment]
ah, it's not worked when i used arrrow function.

![4f222 8o yyabp9jdr958 r](https://cloud.githubusercontent.com/assets/2870324/19461442/23ba2b62-9517-11e6-860d-25ff0a35e9b7.png)

**Comment by Laubeee** [comment]
For anyone ending up here in search of better way (casting really isn't the preferred choice here, which is also why eslint will complain about it), you can either use this
```typescript
app.use(function(err, req, res, next) { ... } as express.ErrorRequestHandler)
```
as suggested by eslint, or you first assign the function to a typed variable, like this:
```typescript
const errHandler: express.ErrorRequestHandler = (err, req, res, next) => { ... };
app.use(errHandler);
```

P.S. I think this can be closed?

**Comment by Hoishin** [comment]
I don't think type assertion or **always having to** use type annotation are ideal solution.

**Comment by noahbrenner** [comment]
I agree with @Hoishin. The docs say that `express` detects error-handling middleware functions based on them having 4 parameters (scroll down to "Error-handling middleware" in [this section](http://expressjs.com/en/4x/api.html#app.use)).

If I understand correctly, that means that a function with 4 parameters will *never* be treated as anything *other* than an error handler, so this sounds like something that TypeScript should be able to define reliably.

**Comment by noahbrenner** [comment]
For anyone in search of a workaround for passing an arrow function: You can wrap the function in another set of parentheses before making the type assertion:

```typescript
// This doesn't work
app.use((err, req, res, next) => { ... } as ErrorRequestHandler);

// But this does
app.use(((err, req, res, next) => { ... }) as ErrorRequestHandler);
```

**Comment by eurostar-fennec-cooper** [comment]
I'm still experiencing this issue with Typescript 3.8.7 (and even on 4.0.2) @types/express: 4.17.4.

After browsing through Express' types, they seem totally correct. Is this fundamentally a Typescript issue?

**Comment by Faris999** [comment]
Hello, I also have the same problem. It doesn't recognize the overload if I don't refactor it to a variable or explicitly typing it.

**Comment by NemoStein** [comment]
Same issue...
VSCode fails to recognize the 4 params version as a `ErrorRequestHandler`
I'm in a javascript project, so the only way to "cast" the function is using jsdoc
```javascript
app.use(/** @type {import('express').ErrorRequestHandler} */ (err, req, res, next) => {
  console.log(err)
})
```
This is horrible!!

---
VSCode 1.57.0
Typescript 4.3.2 & 4.4.0-dev,

**Comment by princefishthrower** [comment]
I've made a pull request in attempt to resolve this issue: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56124

**Comment by staplespeter** [comment]
I have also recently encountered this and would prefer a fix.  I was unable to determine how to correct the typedefs.  Perhaps some of the original authors could help?

@borisyankov 
@micksatana
@samijaber 
@JoseLion 
@dwrss 
@andoshin11
@puneetar
@dfrankland

**Comment by llllvvuu** [comment]
TypeScript can't infer callback arg types unfortunately:

https://github.com/microsoft/TypeScript/issues/47378
https://github.com/microsoft/TypeScript/issues/49020

```typescript
function A(f: (a: string) => void)
function A(f: (a: unknown, b: string) => void)
function A(f: (a: string, b: string) => void) {}

A((a) => void console.log("okay"))
A((a, b) => void console.log("oh no!"))
```

So this will always be an issue with JS libraries that use callbacks, you have to use the workaround provided above.

**Comment by staplespeter** [comment]
Thanks for the explanation @llllvvuu, makes sense.  And also doesn't make sense (in any language) to place an extending parameter before the others, but especially in TypeScript given how it handles overloading.  Does seem like we're stuck with workarounds unless express changes its definitions entirely (seems highly unlikely)
